pkg/app/net/googleoauth.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package net
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"time"
+
+	"github.com/google/cloud-android-orchestration/pkg/app"
+)
+
+// The credentials as sent by the Google OAuth server.
+type RawCredentials struct {
+	AccessTk  string `json:"access_token"`
+	RefreshTk string `json:"refresh_token"`
+	ExpiresIn int    `json:"expires_in"`
+}
+
+// An implementation of the app.Credentials interface.
+type Credentials struct {
+	RawCredentials
+	expiration time.Time
+}
+
+func newCredentials(r RawCredentials) *Credentials {
+	return &Credentials{
+		RawCredentials: r,
+		expiration:     time.Now().Add(time.Duration(r.ExpiresIn) * time.Second),
+	}
+}
+
+func (c *Credentials) AccessToken() string {
+	return c.AccessTk
+}
+
+func (c *Credentials) Expired() bool {
+	return c.expiration.Before(time.Now())
+}
+
+type GoogleOAuthHelper struct {
```

Comment:
Is it possible to use the [Go oauth2 client library](https://github.com/googleapis/google-api-go-client/blob/main/GettingStarted.md) here? The [official Google APIs OAuth page](https://developers.google.com/identity/protocols/oauth2/web-server) doesn't have code examples for it inline but it does mention it at the top.

Reply:
Unless I missed something (wouldn't be the first time) those libraries are for accessing the Google APIs, but don't have (convenient) utilities to manage the oauth credentials. In this case we need the latter and the former is left to `fetch_cvd`.
Even if those libraries do support our use case I couldn't find any good documentation and examples, so it was just easier to use http calls directly.

Reply:
From the [docs](https://pkg.go.dev/golang.org/x/oauth2), which includes a [code sample](https://go.dev/play/p/4VQcYjHb6tG):

- `GoogleOAuthHelper` is [`oauth2.Config`](https://pkg.go.dev/golang.org/x/oauth2#Config)
- `RawCredentials` is [`oauth2.Token`](https://pkg.go.dev/golang.org/x/oauth2#Token)
- `AuthorizationURL` is [`AuthCodeURL`](https://pkg.go.dev/golang.org/x/oauth2#Config.AuthCodeURL)
- `ParseAuthorizationResponse` is [`Exchange`](https://pkg.go.dev/golang.org/x/oauth2#Config.Exchange)
- `RefreshCredentials` maps to either [`TokenSource`](https://pkg.go.dev/golang.org/x/oauth2#Config.TokenSource) or [`Client`](https://pkg.go.dev/golang.org/x/oauth2#Config.Client)

Reply:
Sorry, I should have lead with a link to the [oauth2 go package](https://pkg.go.dev/golang.org/x/oauth2) which is referenced there, but is more relevant than the `google-api-go-client` package.

Reply:
Thanks! Rewrote it with that library instead.